File: src/main/java/com/gitlab/lae/intellij/actions/tree/When.java
package com.gitlab.lae.intellij.actions.tree;

import com.google.auto.value.AutoValue;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.wm.IdeFocusManager;
import com.intellij.openapi.wm.ToolWindow;

import javax.swing.text.JTextComponent;
import java.awt.*;
import java.util.List;
import java.util.function.Predicate;
import java.util.regex.Pattern;

import static com.intellij.openapi.actionSystem.CommonDataKeys.VIRTUAL_FILE;
import static com.intellij.openapi.actionSystem.PlatformDataKeys.TOOL_WINDOW;
import static java.util.Arrays.asList;
import static java.util.Collections.unmodifiableList;

public abstract class When implements Predicate<DataContext> {
    When() {
    }

    public static final When ALWAYS = new When() {
        @Override
        public boolean test(DataContext context) {
            return true;
        }

        @Override
        public String toString() {
            return "When.ALWAYS";
        }
    };

    public static final When NEVER = new When() {
        @Override
        public boolean test(DataContext context) {
            return false;
        }

        @Override
        public String toString() {
            return "When.NEVER";
        }
    };

    static final When INPUT_FOCUSED = new When() {
        @Override
        public boolean test(DataContext context) {
            IdeFocusManager focusManager =
                    IdeFocusManager.findInstanceByContext(context);
            Component component = focusManager.getFocusOwner();
            return component instanceof JTextComponent &&
                    ((JTextComponent) component).isEditable();
        }

        @Override
        public String toString() {
            return "When.INPUT_FOCUSED";
        }
    };

    public static When parse(String input) {
        return input.startsWith("!")
                ? When.not(doParse(input.substring(1)))
                : doParse(input);
    }

    private static When doParse(String input) {
        if (input.equals("InputFocused")) {
            return INPUT_FOCUSED;
        }
        String[] parts = input.split(":", 2);
        if (parts.length != 2) {
            throw new IllegalArgumentException(
                    "Invalid 'when' pattern: '" + input + "'");
        }
        String type = parts[0];
        String arg = parts[1];
        switch (type) {
            case "ToolWindowActive": return toolWindowActive(arg);
            case "ToolWindowTabActive": return toolWindowTabActive(arg);
            case "FileExtension": return fileExtension(arg);
        }
        throw new IllegalArgumentException(input);
    }

    public static When any(When... clauses) {
        return new AutoValue_When_Any(unmodifiableList(asList(clauses)));
    }

    public static When all(When... clauses) {
        return new AutoValue_When_All(unmodifiableList(asList(clauses)));
    }

    public static When toolWindowActive(String titleRegex) {
        return new AutoValue_When_ToolWindowActive(Pattern.compile(titleRegex));
    }

    public static When toolWindowTabActive(String tabTitleRegex) {
        return new AutoValue_When_ToolWindowTabActive(
                Pattern.compile(tabTitleRegex));
    }

    public static When fileExtension(String extRegex) {
        return new AutoValue_When_FileExtension(Pattern.compile(extRegex));
    }

    public static When not(When when) {
        return new AutoValue_When_Not(when);
    }

    @AutoValue
    static abstract class Any extends When {
        abstract List<When> clauses();

        @Override
        public boolean test(DataContext context) {
            return clauses().stream().anyMatch(p -> p.test(context));
        }
    }

    @AutoValue
    static abstract class All extends When {
        abstract List<When> clauses();

        @Override
        public boolean test(DataContext context) {
            return clauses().stream().allMatch(p -> p.test(context));
        }
    }

    @AutoValue
    static abstract class Not extends When {
        abstract When when();

        @Override
        public boolean test(DataContext context) {
            return !when().test(context);
        }
    }

    static abstract class Regex extends When {

        abstract Pattern regex();

        abstract String value(DataContext context);

        @Override
        public boolean test(DataContext context) {
            String value = value(context);
            if (value == null) {
                return false;
            }
            return regex().matcher(value).matches();
        }

        @Override
        public int hashCode() {
            return regex().pattern().hashCode();
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == null) {
                return false;
            }
            if (obj.getClass() != getClass()) {
                return false;
            }
            return regex().pattern().equals(((Regex) obj).regex().pattern());
        }
    }

    @AutoValue
    static abstract class ToolWindowActive extends Regex {
        @Override
        String value(DataContext context) {
            ToolWindow window = context.getData(TOOL_WINDOW);
            return window != null && window.isActive()
                    ? window.getStripeTitle()
                    : null;
        }
    }

    @AutoValue
    static abstract class ToolWindowTabActive extends Regex {
        @Override
        String value(DataContext context) {
            ToolWindow window = context.getData(TOOL_WINDOW);
            return window != null && window.isActive()
                    ? window.getTitle()
                    : null;
        }
    }

    @AutoValue
    static abstract class FileExtension extends Regex {
        @Override
        String value(DataContext context) {
            VirtualFile file = context.getData(VIRTUAL_FILE);
            return file != null ? file.getExtension() : null;
        }
    }
}

File: src/main/kotlin/com/gitlab/lae/intellij/actions/tree/When.kt
package com.gitlab.lae.intellij.actions.tree

import com.intellij.openapi.actionSystem.CommonDataKeys.PROJECT
import com.intellij.openapi.actionSystem.CommonDataKeys.VIRTUAL_FILE
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.actionSystem.DataProvider
import com.intellij.openapi.actionSystem.PlatformDataKeys.EDITOR
import com.intellij.openapi.actionSystem.PlatformDataKeys.TOOL_WINDOW
import com.intellij.openapi.wm.IdeFocusManager
import java.nio.file.Files.exists
import java.nio.file.Path
import java.nio.file.Paths
import java.util.function.Predicate
import java.util.regex.Pattern
import javax.swing.text.JTextComponent

interface When : Predicate<DataContext> {

  data class Any(val clauses: List<When>) : When {

    constructor(vararg clauses: When) :
      this(listOf(*clauses))

    override fun test(context: DataContext) =
      clauses.any { it.test(context) }
  }

  data class All(val clauses: List<When>) : When {

    constructor(vararg clauses: When) :
      this(listOf(*clauses))

    override fun test(context: DataContext) =
      clauses.all { it.test(context) }
  }

  data class Not(val condition: When) : When {
    override fun test(context: DataContext) =
      !condition.test(context)
  }

  class EqPattern(val pattern: Pattern) {
    constructor(pattern: String) : this(Pattern.compile(pattern))

    override fun toString() = pattern.toString()
    override fun hashCode() = pattern.hashCode()
    override fun equals(other: kotlin.Any?) =
      other is EqPattern && pattern.pattern() == other.pattern.pattern()
  }

  interface Regex : When {
    val regex: EqPattern
    fun value(context: DataContext): String?
    override fun test(context: DataContext): Boolean {
      val value = value(context) ?: return false
      return regex.pattern.matcher(value).matches()
    }
  }

  data class ToolWindowActive(override val regex: EqPattern) : Regex {

    constructor(regex: String) :
      this(EqPattern(regex))

    override fun value(context: DataContext): String? {
      val window = context.getData(TOOL_WINDOW)
      return if (window != null && window.isActive) window.stripeTitle else null
    }
  }

  data class ToolWindowTabActive(override val regex: EqPattern) : Regex {

    constructor(regex: String) :
      this(EqPattern(regex))

    override fun value(context: DataContext): String? {
      val window = context.getData(TOOL_WINDOW)
      return if (window != null && window.isActive) window.title else null
    }
  }

  data class FileExtension(override val regex: EqPattern) : Regex {

    constructor(regex: String) :
      this(EqPattern(regex))

    override fun value(context: DataContext) =
      context.getData(VIRTUAL_FILE)?.extension
  }

  data class PathExists(val path: Path) : When {

    constructor(path: String) :
      this(Paths.get(path))

    override fun test(context: DataContext) =
      if (path.isAbsolute) {
        exists(path)
      } else {
        context.getData(PROJECT)?.basePath?.let {
          exists(Paths.get(it).resolve(path))
        } ?: false
      }
  }

  object Always : When {
    override fun toString() = "When.Always"
    override fun test(context: DataContext) = true
  }

  object Never : When {
    override fun toString() = "When.Never"
    override fun test(context: DataContext) = false
  }

  object InputFocused : When {
    override fun toString() = "When.InputFocused"
    override fun test(context: DataContext): Boolean {
      val focusManager = IdeFocusManager.findInstanceByContext(context)
      val component = focusManager.focusOwner
      return component is JTextComponent && component.isEditable
    }
  }

  object TextSelected : When {
    override fun toString() = "When.TextSelected"

    override fun test(context: DataContext): Boolean {
      val focusManager = IdeFocusManager.findInstanceByContext(context)
      val component = focusManager.focusOwner
      if (component !is JTextComponent) {
        return false
      }

      val hasComponentSelection =
        component.selectionStart < component.selectionEnd

      val hasEditorSelection =
        component is DataProvider
          && EDITOR.getData(component)?.selectionModel?.hasSelection(true)
          ?: false

      return hasComponentSelection || hasEditorSelection;
    }
  }

  companion object {

    fun parse(input: String): When =
      if (input.startsWith("!")) Not(doParse(input.substring(1)))
      else doParse(input)

    private fun doParse(input: String): When {
      when (input) {
        "InputFocused" -> return InputFocused
        "TextSelected" -> return TextSelected
      }

      val parts = input.split(":".toRegex(), 2).toTypedArray()
      require(parts.size == 2) { "Invalid 'when' pattern: '$input'" }

      val type = parts[0]
      val arg = parts[1]
      return when (type) {
        "ToolWindowActive" -> ToolWindowActive(arg)
        "ToolWindowTabActive" -> ToolWindowTabActive(arg)
        "FileExtension" -> FileExtension(arg)
        "PathExists" -> PathExists(arg)
        else -> throw IllegalArgumentException(input)
      }
    }
  }
}
